src/HasEnums.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace Spatie\Enum;
+
+use Spatie\Enum\Exceptions\InvalidEnumError;
+use Spatie\Enum\Exceptions\InvalidValueException;
+
+trait HasEnums
+{
+    /**
+     * @param $key
+     * @param \Spatie\Enum\Enum $enumObject
+     *
+     * @return mixed
+     */
+    public function setAttribute($key, $enumObject)
+    {
+        if (! isset($this->enums[$key])) {
+            return parent::setAttribute($key, $enumObject);
+        }
+
+        $enumClass = $this->enums[$key];
+
+        if (! is_a($enumObject, $enumClass)) {
+            throw InvalidEnumError::make(
+                static::class,
+                $key,
+                $enumClass,
+                get_class($enumObject)
+            );
+        }
+
+        $enumValue = $enumObject->getValue();
+
+        $mappedValue = $enumClass::$map[$enumValue] ?? null;
```

Comment:
Instead of a whole new map I would like to define if you want the enum `value` or the `index` as attribute value and use corresponding `getValue()` or `getIndex()` methods.
This would remove duplicate mapping and keep it simple.